File: RPN.java
import java.util.List;
import java.util.Stack;
import java.util.LinkedList;

public class RPN {

  static void p(String s) { System.out.println(s); }

  // @include
  static boolean isOperator(char c) {
    return (c == '*' || c == '+' || c == '-' || c == '/');
  }

  public static int eval(String s) {
     Stack<Integer> evalStack = new Stack<Integer>();
     int offset = 0, length = s.length();

     while (offset < length) {
       char c;
       while (offset < length && isOperator(c = s.charAt(offset))) {
         Integer x, y;
         x = evalStack.pop();
         y = evalStack.pop();
         switch (c) {
           case '*':
             evalStack.push(x * y); break;
           case '+':
             evalStack.push(x + y); break;
           case '/':
             evalStack.push(x / y); break;
           case '-':
             evalStack.push(x - y); break;
           default:
             break;
         }
         ++offset;
       }

       if (offset >= length)
         break;

       int numIndex = offset;
       while (Character.isDigit(s.charAt(offset)))
         ++offset;
       
       Integer tmp = Integer.parseInt(s.substring(numIndex, offset));
       // Detect negative number
       if (s.charAt(offset) == '-') {
         tmp = -tmp;
         ++offset;
       }
       if (s.charAt(offset) == ',') {
         ++offset;
       }
       evalStack.push(tmp);
     }
     return evalStack.pop();
  }
  // @exclude

  static public void main(String [] args) {
    System.out.println(eval("2,10-,/"));
    assert (-5 == eval("2,10-,/"));
    System.out.println(eval("10,2+"));
    assert (12 == eval("10,2+"));
    System.out.println(eval("1,2+3,4*+"));
    assert (15 == eval("1,2+3,4*+"));
    System.out.println(eval("1,2,3,4,5+*++3,4*+"));
    assert (42 == eval("1,2,3,4,5+*++3,4*+"));
  }
}

File: RPN.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <sstream>
#include <stack>
#include <string>

using std::stack;
using std::string;
using std::stringstream;

// @include
int eval(const string& s) {
  stack<int> eval_stack;
  stringstream ss(s);
  string symbol;

  while (getline(ss, symbol, ',')) {
    if (symbol == "+" || symbol == "-" || symbol == "*" || symbol == "/") {
      int y = eval_stack.top();
      eval_stack.pop();
      int x = eval_stack.top();
      eval_stack.pop();
      switch (symbol.front()) {
        case '+':
          eval_stack.emplace(x + y);
          break;
        case '-':
          eval_stack.emplace(x - y);
          break;
        case '*':
          eval_stack.emplace(x * y);
          break;
        case '/':
          eval_stack.emplace(x / y);
          break;
      }
    } else {  // number.
      eval_stack.emplace(stoi(symbol));
    }
  }
  return eval_stack.top();
}
// @exclude

int main(int argc, char* argv[]) {
  assert(0 == eval("2,-10,/"));
  assert(-5 == eval("-10,2,/"));
  assert(5 == eval("-10,-2,/"));
  assert(-5 == eval("5,10,-"));
  assert(6 == eval("-10,-16,-"));
  assert(12 == eval("10,2,+"));
  assert(15 == eval("1,2,+,3,4,*,+"));
  assert(42 == eval("1,2,3,4,5,+,*,+,+,3,4,*,+"));
  return 0;
}
